File: tools/parse_call.py
"""
parse_call.py
Uses the Anthropic API to parse a call transcript into structured compliance JSON.

Usage:
    python tools/parse_call.py --log_id <call_log_uuid> [--dry_run true]

Reads the transcript from call_logs, parses it, and writes parsed_results back.
"""

import argparse
import json
import os

import anthropic
from dotenv import load_dotenv
from supabase import create_client

load_dotenv()

SUPABASE_URL = os.environ["SUPABASE_URL"]
SUPABASE_SERVICE_KEY = os.environ["SUPABASE_SERVICE_KEY"]
ANTHROPIC_API_KEY = os.environ["ANTHROPIC_API_KEY"]

SYSTEM_PROMPT = """You are a clinical compliance parser for a cardiac rehabilitation programme.
Given a call transcript between an AI and an elderly cardiac patient, extract structured compliance data.

Return ONLY valid JSON — an array of task results:
[
  {
    "task_name": "<task identifier>",
    "completed": true | false | null,
    "notes": "<brief free-text note — max 100 chars>",
    "value_reported": "<optional — e.g. heart rate '88 bpm', walk duration '15 minutes'>",
    "alert_flag": "normal" | "watch" | "urgent"
  }
]

Rules:
- completed = true if patient confirmed doing the task
- completed = false if patient said they didn't do it or it's clear they missed it
- completed = null if unclear or not discussed
- alert_flag = "urgent" if ANY cardiac symptom: chest pain, pressure, shortness of breath at rest, near-fainting, severe confusion
- alert_flag = "watch" if: missed 2+ meds, ankle swelling, poor sleep 3+ nights, exercise missed 2+ days
- alert_flag = "normal" otherwise
- If the patient went off-topic, note that but still extract what you can.
- Respond in the same language as the transcript."""


def parse_call(log_id: str, dry_run: bool = False) -> list:
    supabase = create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)

    log = supabase.table("call_logs").select("*").eq("id", log_id).single().execute()
    if not log.data:
        raise ValueError(f"Call log {log_id} not found")
    entry = log.data

    if not entry.get("transcript"):
        raise ValueError(f"No transcript found for log {log_id}")

    # Fetch the tasks for context
    patient_id = entry["patient_id"]
    call_slot = entry["call_type"]
    plan = (
        supabase.table("recovery_plans")
        .select("id")
        .eq("patient_id", patient_id)
        .order("created_at", desc=True)
        .limit(1)
        .execute()
    )
    tasks = []
    if plan.data:
        tasks_result = (
            supabase.table("tasks")
            .select("task_name, task_type, is_alert_trigger")
            .eq("plan_id", plan.data[0]["id"])
            .eq("call_slot", call_slot)
            .execute()
        )
        tasks = tasks_result.data or []

    task_context = "\n".join(f"- {t['task_name']} (alert_trigger: {t['is_alert_trigger']})" for t in tasks)
    user_prompt = f"""Call slot: {call_slot}
Expected tasks:
{task_context if task_context else '(no specific tasks — infer from transcript)'}

Transcript:
{entry['transcript']}

Parse the compliance results now."""

    if dry_run:
        print("[DRY RUN] Would call Anthropic API with transcript:")
        print(entry["transcript"][:200], "...")
        return [{"dry_run": True}]

    client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)
    message = client.messages.create(
        model="claude-opus-4-6",
        max_tokens=1024,
        system=SYSTEM_PROMPT,
        messages=[{"role": "user", "content": user_prompt}],
    )

    raw = message.content[0].text.strip()
    if raw.startswith("```"):
        raw = raw.split("```")[1]
        if raw.startswith("json"):
            raw = raw[4:]
    parsed = json.loads(raw)

    # Determine overall severity flag
    flags = [r.get("alert_flag", "normal") for r in parsed]
    if "urgent" in flags:
        overall_flag = "urgent"
    elif "watch" in flags:
        overall_flag = "watch"
    else:
        overall_flag = "normal"

    # Write results back to call_logs
    supabase.table("call_logs").update({
        "parsed_results": parsed,
        "severity_flag": overall_flag,
        "status": "completed",
    }).eq("id", log_id).execute()

    print(json.dumps(parsed, indent=2, ensure_ascii=False))
    return parsed


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--log_id", required=True)
    parser.add_argument("--dry_run", default="false")
    args = parser.parse_args()

    dry = args.dry_run.lower() == "true"
    parse_call(args.log_id, dry_run=dry)

File: lib/tools/parseCall.ts
import Anthropic from "@anthropic-ai/sdk";
import { createServerClient } from "@/lib/supabase";

const SYSTEM_PROMPT = `You are a clinical compliance parser for a cardiac rehabilitation programme.
Given a call transcript between an AI and an elderly cardiac patient, extract structured compliance data.

Return ONLY valid JSON — an array of task results:
[
  {
    "task_name": "<task identifier>",
    "completed": true | false | null,
    "notes": "<brief free-text note — max 100 chars>",
    "value_reported": "<optional — e.g. heart rate '88 bpm', or the raw rating e.g. '7/10'>",
    "rating": <1-10 integer if the patient gave a numeric rating, otherwise null>,
    "conclusion": "<human-readable interpretation combining the rating AND the patient's own description — e.g. 'Sleep rated 7/10 - Slept well but woke up once' or 'Chest pain rated 3/10 - Mild tightness in the morning'>",
    "alert_flag": "normal" | "watch" | "urgent"
  }
]

Rules:
- completed = true if patient confirmed doing the task or gave a rating indicating acceptable status
- completed = false if patient said they didn't do it or the rating indicates a significant problem
- completed = null if unclear or not discussed
- Transcript lines marked [low confidence: X.XX] or [conf: X.XX] — treat as completed = null unless clearly interpretable

Rating interpretation guidelines (use for conclusion and alert_flag):
- Sleep quality (higher = better): 1-3 Poor (watch), 4-6 Fair (normal), 7-10 Good (normal)
- Chest pain/discomfort (higher = worse): 1-2 None/minimal (normal), 3-5 Mild (watch), 6-8 Moderate (urgent), 9-10 Severe (urgent)
- Breathlessness (higher = worse): 1-3 Minimal (normal), 4-6 Moderate (watch), 7-10 Severe (urgent)
- Fatigue/energy (higher = better): 1-3 Very fatigued (watch), 4-6 Moderate (normal), 7-10 Good energy (normal)
- Overall wellbeing (higher = better): 1-3 Poor (watch), 4-6 Moderate (normal), 7-10 Good (normal)
- Swelling/oedema (higher = worse): 1-3 Minimal (normal), 4-6 Moderate (watch), 7-10 Severe (urgent)

- alert_flag = "urgent" if: chest pain rated 6+, breathlessness rated 7+, swelling rated 7+, or any direct cardiac symptom mentioned
- alert_flag = "watch" if: sleep rated 1-3, fatigue rated 1-3, chest pain rated 3-5, breathlessness rated 4-6, missed 2+ medications
- alert_flag = "normal" otherwise`;

export async function parseCall(logId: string, dryRun = false): Promise<object[]> {
  const supabase = createServerClient();

  const { data: log, error } = await supabase
    .from("call_logs")
    .select("*")
    .eq("id", logId)
    .single();
  if (error || !log) throw new Error(`Call log ${logId} not found`);
  if (!log.transcript) throw new Error(`No transcript for log ${logId}`);

  const { data: plans } = await supabase
    .from("recovery_plans")
    .select("id")
    .eq("patient_id", log.patient_id)
    .order("created_at", { ascending: false })
    .limit(1);

  let taskContext = "";
  if (plans && plans.length > 0) {
    const { data: tasks } = await supabase
      .from("tasks")
      .select("task_name, is_alert_trigger")
      .eq("plan_id", plans[0].id)
      .eq("call_slot", log.call_type);
    if (tasks) {
      taskContext = tasks.map((t) => `- ${t.task_name} (alert_trigger: ${t.is_alert_trigger})`).join("\n");
    }
  }

  if (dryRun) {
    return [{ dry_run: true, transcript_length: log.transcript.length }];
  }

  const userPrompt = `Call slot: ${log.call_type}
Expected tasks:
${taskContext || "(no specific tasks — infer from transcript)"}

Transcript:
${log.transcript}

Parse the compliance results now.`;

  const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
  const message = await client.messages.create({
    model: "claude-opus-4-6",
    max_tokens: 1024,
    system: SYSTEM_PROMPT,
    messages: [{ role: "user", content: userPrompt }],
  });

  let raw = (message.content[0] as { type: string; text: string }).text.trim();
  if (raw.startsWith("```")) {
    raw = raw.split("```")[1];
    if (raw.startsWith("json")) raw = raw.slice(4);
  }
  const parsed: { alert_flag: string }[] = JSON.parse(raw);

  const flags = parsed.map((r) => r.alert_flag);
  const overallFlag = flags.includes("urgent") ? "urgent" : flags.includes("watch") ? "watch" : "normal";

  await supabase
    .from("call_logs")
    .update({ parsed_results: parsed, severity_flag: overallFlag, status: "completed" })
    .eq("id", logId);

  return parsed;
}
